File: 1675-magnetic-force-between-two-balls/1675-magnetic-force-between-two-balls.java
class Solution {
    public int maxDistance(int[] position, int m) {
        Arrays.sort(position);
        int min = 1;
        int max =position[position.length - 1] - position[0];
        int mid = 0;
        int ans=0;
        while(min <= max){
            mid =(min + max)/2;
            if(isPossible(position, mid, m)){
                ans=mid;
                min = mid + 1;
            }
            else{
                max = mid - 1;
            }
        }
        return ans;

    }
    public boolean isPossible(int[] position, int mid, int m){
        m= m-1;
        int j=0;
        for(int i=1;i<position.length;i++){
            if(position[i] - position[j] >= mid){
                m--;
                j=i;
            }
        }
        return m<=0;
    }
}

File: 1675-magnetic-force-between-two-balls/1675-magnetic-force-between-two-balls.py
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:

        def isPossible(position: List[int], m: int,mid: int):
            count=1
            j=position[0]
            for i in range(1,len(position)):
                force=position[i]-j
                if(force>=mid):
                    count += 1
                    j=position[i]
            return count>=m
        position.sort()
        min_v=1
        max_v=position[-1]-position[0]
        answer=0
        while(min_v<=max_v):
            mid=(min_v+max_v)//2
            if(isPossible(position,m,mid)):
                answer=mid
                min_v=mid+1
            else:
                max_v=mid-1
        return answer